### PR TITLE
Handle ShouldRunUpdate in managed code

### PR DESCRIPTION
**Pull request by jechter**
Following https://ono.unity3d.com/unity/unity/pull-request/83610/_/platform/foundation/input-should-update, we can now move the logic to decide if we want to run updates to managed code. This PR does that.

### CODE REVIEW

File: Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs [hunk 1]
```diff
@@ -151,6 +151,19 @@ public Action<InputUpdateType> onBeforeUpdate
             }
```
Comment (douglas-piconi) [inline Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs:151]: Im getting a compile error on this branch, NativeInputRuntime.cs(16,41): error CS0535: 'NativeInputRuntime' does not implement interface member 'IInputRuntime.runInBackground'

Comment (douglas-piconi) [reply Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs:151]: Do we have to land any native first?

Comment (jechter) [reply Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs:151]: Good catch. Seems I committed incomplete changes (guess that is me being too used to hg workflows where last-minute changes to "staged" files will also be staged, whereas in git changes are "staged" on a per-change level, not per-file).

Fixed.

File: Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs [hunk 2]
```diff
@@ -151,6 +151,19 @@ public Action<InputUpdateType> onBeforeUpdate
...
+        public Func<InputUpdateType, bool> onShouldRunUpdate
+        {
+            set
+            {
+                // This is stupid but the enum prevents us from jacking the delegate in directly.
+                // This means we get a double dispatch here :(
+                if (value != null)
+                    NativeInputSystem.onShouldRunUpdate = updateType => value((InputUpdateType)updateType);
+                else
+                    NativeInputSystem.onShouldRunUpdate = null;
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs:163]: Are the native changes for this already in 2019.1?

File: Packages/com.unity.inputsystem/Tests/InputSystem/CoreTests_Events.cs [hunk 3]
```diff
@@ -237,6 +237,67 @@ public void Events_CanSwitchToProcessingInFixedUpdatesOnly()
...
+        Assert.That(runtime.onShouldRunUpdate.Invoke(InputUpdateType.Dynamic), Is.EqualTo(runInBackground));
+        Assert.That(runtime.onShouldRunUpdate.Invoke(InputUpdateType.Fixed), Is.EqualTo(runInBackground));
+        Assert.That(runtime.onShouldRunUpdate.Invoke(InputUpdateType.Editor), Is.EqualTo(!runInBackground));
+
+        UnityEditor.EditorApplication.ExecuteMenuItem("Window/General/Game");
+
+        Assert.That(runtime.onShouldRunUpdate.Invoke(InputUpdateType.Dynamic));
+        Assert.That(runtime.onShouldRunUpdate.Invoke(InputUpdateType.Fixed));
+        Assert.That(!runtime.onShouldRunUpdate.Invoke(InputUpdateType.Editor));
+    }
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/Tests/InputSystem/CoreTests_Events.cs:271]: Relying on window focus in a play mode test seems brittle. I think it's better to just test this at the `IInputRuntime` level. I.e. have `InputTestRuntime` go through focus changes and make sure the input system reacts properly.

**Review by Rene-Damm — COMMENTED** [review]
Crap, again got caught by forgetting to hit "Submit" again to, like, actually post comments...

File: Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs (hunk 2 above)
Comment (jechter) [reply Packages/com.unity.inputsystem/InputSystem/NativeInputRuntime.cs:163]: yes.

File: Packages/com.unity.inputsystem/Tests/InputSystem/CoreTests_Events.cs (hunk 3 above)
Comment (jechter) [reply Packages/com.unity.inputsystem/Tests/InputSystem/CoreTests_Events.cs:271]: I agree that this is brittle, yeah. So you mean expose focus as a property of IInputRuntime instead (so we can override it in tests)? Yeah, I guess that makes sense.

**Pull request merged**